File: java/Automata.java
final class Automata {
    static int[][] kernelFromMoore8(int[] coeffs) {
        return new int[][] {
            { coeffs[0], coeffs[1], coeffs[2] },
            { coeffs[7], 0, coeffs[3] },
            { coeffs[6], coeffs[5], coeffs[4] }
        };
    }

    static void stepU16(
        short[] prev,
        short[] curPerm,
        short[] next,
        int height,
        int width,
        int channels,
        int[][] k1,
        int[][] k2,
        int round,
        Common.BoundaryConfig bc
    ) {
        for (int y = 0; y < height; y++) {
            for (int x = 0; x < width; x++) {
                for (int c = 0; c < channels; c++) {
                    long acc = 0;
                    int[][] k = (((y + x + c + round) & 1) == 0) ? k1 : k2;
                    for (int dy = -1; dy <= 1; dy++) {
                        for (int dx = -1; dx <= 1; dx++) {
                            int ny = mapY(y + dy, height, bc);
                            int nx = mapX(x + dx, width, bc);
                            int idx = ((ny * width) + nx) * channels + c;
                            acc += (long)k[dy + 1][dx + 1] * Short.toUnsignedInt(curPerm[idx]);
                        }
                    }
                    int pos = ((y * width) + x) * channels + c;
                    int conv = mod257(acc);
                    next[pos] = (short)mod257(conv - Short.toUnsignedInt(prev[pos]));
                }
            }
        }
    }

    private static int mod257(long x) {
        long r = x % Common.AUTOMATA_MOD_BASE;
        return (int)(r < 0 ? r + Common.AUTOMATA_MOD_BASE : r);
    }

    private static int clamp(int idx, int max) {
        return idx < 0 ? 0 : (idx >= max ? max - 1 : idx);
    }

    private static int wrap(int idx, int max) {
        int r = idx % max;
        return r < 0 ? r + max : r;
    }

    private static int mapY(int y, int h, Common.BoundaryConfig bc) {
        if (y >= 0 && y < h) return y;
        if (y < 0) return bc.top() == Common.BoundaryMode.PERIODIC ? wrap(y, h) : clamp(y, h);
        return bc.bottom() == Common.BoundaryMode.PERIODIC ? wrap(y, h) : clamp(y, h);
    }

    private static int mapX(int x, int w, Common.BoundaryConfig bc) {
        if (x >= 0 && x < w) return x;
        if (x < 0) return bc.left() == Common.BoundaryMode.PERIODIC ? wrap(x, w) : clamp(x, w);
        return bc.right() == Common.BoundaryMode.PERIODIC ? wrap(x, w) : clamp(x, w);
    }
}

File: cs/Automata.cs
namespace TTCrypto;

public static class Automata
{
    public static int[,] KernelFromMoore8(ushort[] coeffs)
    {
        return new[,]
        {
            { coeffs[0], coeffs[1], coeffs[2] },
            { coeffs[7], 0, coeffs[3] },
            { coeffs[6], coeffs[5], coeffs[4] }
        };
    }

    public static void StepU16(
        ushort[] prevState,
        ushort[] curPermuted,
        ushort[] nextState,
        uint height,
        uint width,
        ushort channels,
        int[][,] kernelsByChannel,
        uint roundIndex,
        BoundaryConfig bc)
    {
        for (uint y = 0; y < height; y++)
        {
            for (uint x = 0; x < width; x++)
            {
                for (ushort c = 0; c < channels; c++)
                {
                    long acc = 0;
                    var kernel = SelectKernel(y, x, c, roundIndex, kernelsByChannel);
                    for (var dy = -1; dy <= 1; dy++)
                    {
                        for (var dx = -1; dx <= 1; dx++)
                        {
                            var ny = MapY((int)y + dy, (int)height, bc);
                            var nx = MapX((int)x + dx, (int)width, bc);
                            var idx = Index3((uint)ny, (uint)nx, c, width, channels);
                            acc += (long)kernel[dy + 1, dx + 1] * curPermuted[idx];
                        }
                    }
                    var pos = Index3(y, x, c, width, channels);
                    var conv = Mod257(acc);
                    nextState[pos] = (ushort)Mod257(conv - prevState[pos]);
                }
            }
        }
    }

    private static int[,] SelectKernel(uint y, uint x, ushort c, uint roundIndex, int[][,] kernelsByChannel)
        => kernelsByChannel[(c % 3) * 2 + ((((y + x + c + roundIndex) & 1u) == 0u) ? 0 : 1)];

    private static int Index3(uint y, uint x, ushort c, uint width, ushort channels)
        => ((int)y * (int)width + (int)x) * channels + c;

    private static int Mod257(long x)
    {
        var r = x % CryptoConsts.AutomataModBase;
        if (r < 0) r += CryptoConsts.AutomataModBase;
        return (int)r;
    }

    private static int Clamp(int idx, int max) => idx < 0 ? 0 : (idx >= max ? max - 1 : idx);

    private static int Wrap(int idx, int max)
    {
        var r = idx % max;
        return r < 0 ? r + max : r;
    }

    private static int MapY(int y, int h, BoundaryConfig bc)
        => y >= 0 && y < h ? y : (y < 0 ? (bc.Top == BoundaryMode.Periodic ? Wrap(y, h) : Clamp(y, h))
                                        : (bc.Bottom == BoundaryMode.Periodic ? Wrap(y, h) : Clamp(y, h)));

    private static int MapX(int x, int w, BoundaryConfig bc)
        => x >= 0 && x < w ? x : (x < 0 ? (bc.Left == BoundaryMode.Periodic ? Wrap(x, w) : Clamp(x, w))
                                        : (bc.Right == BoundaryMode.Periodic ? Wrap(x, w) : Clamp(x, w)));
}
